[NVPTX] add combiner rule for final packed op in reduction

Prince781 opened the pull request:
For vector reductions, the final result needs to be a scalar. The default
expansion will use packed ops (ex. `fadd.f16x2`) even for the final operation.
This requires a packed operation where one of the lanes is `undef`.
                                                                             
ex: lowering of `vecreduce_fadd(V)` where `V = v4f16<a b c d>`

```
v1: v2f16 = fadd reassoc v2f16<a b>, v2f16<c d>  (== <a+c     b+d>)
v2: v2f16 = vector_shuffle<1,u> v1, undef:v2f16  (== <b+d     undef>)
v3: v2f16 = fadd reassoc v2, v1                  (== <b+d+a+c undef>)
vR: f16   = extractelt v3, 1
```
                                                                             
We wish to replace `vR`, `v3`, and `v2` with:

```
vR: f16 = fadd reassoc (extractelt v1, 1) (extractelt v1, 0)
```
                                                                             
...so that we get:

```
v1: v2f16 = fadd reassoc v2f16<a b>, v2f16<c d>  (== <a+c     b+d>)
s1: f16   = extractelt v1, 1
s2: f16   = extractelt v1, 0
vR: f16   = fadd reassoc s1, s2                  (== b+d+a+c)
```

So for this example, this rule will replace `v3` and `v2`, returning a vector
with the result in lane 0 and an undef in lane 1, which we expect will be
folded into the `extractelt` in `vR`.

Prince781 committed:
[NVPTX] add combiner rule for final packed op in reduction

justinfargnoli reviewed (commented):
This seems to be a fix for an inefficient lowering of `vecreduce_fadd`.

NVPTX isn't custom lowering `vecreduce_fadd`, correct? If that is the case, then this is an issue other targets could/do face. 

Given that, why not improve the `vecreduce_fadd` lowering or implement this in DAGCombiner?

Prince781 committed:
Revert "[NVPTX] add combiner rule for final packed op in reduction"

This reverts commit 8cbda008607cbcdd8df3a0cc5994349d81fc81b5.

Prince781 committed:
[DAGCombiner] add rule for vector reduction op with undef lane

The result of a reduction needs to be a scalar. When expressed as a
sequence of vector ops, the final op needs to reduce two (or more)
lanes belonging to the same register. On some targets a shuffle is not
available and this results in two extra movs to setup another vector
register with the lanes swapped. This pattern is now handled better by
turning it into a scalar op.

Prince781 commented:
I ported this to DAGCombiner, but the optimization only winds up benefiting NVPTX. What do you guys think? Should we keep it in NVPTXISelLowering?

Prince781 commented:
Some targets like x86 seem to handle this sequence just fine.

Prince781 committed:
Revert "[DAGCombiner] add rule for vector reduction op with undef lane"

This reverts commit 371a3ad352b19ff42190e64f0360d9f79245e25f.

Prince781 committed:
Reapply "[NVPTX] add combiner rule for final packed op in reduction"

This reverts commit 9d84bd34f53646a6769331d40516e150a3e05bf2.

Prince781 commented:
So I've decided to merge this back into NVPTXISelLowering.

Prince781 commented:
Superseded by #136253
